src/Nonlinear/ReverseAD/Coloring/Coloring.jl

```diff
@@ -0,0 +1,547 @@
+module Coloring
+
+import DataStructures
+
+include("topological_sort.jl")
+
+# indexed sparse set of integers
+mutable struct IndexedSet
+    nzidx::Vector{Int}
+    empty::BitArray{1}
+    nnz::Int
+end
+
+IndexedSet(n::Integer) = IndexedSet(zeros(Int, n), trues(n), 0)
+
+function Base.push!(v::IndexedSet, i::Integer)
+    if v.empty[i]  # new index
+        v.nzidx[v.nnz+=1] = i
+        v.empty[i] = false
+    end
+    return
+end
+
+function Base.empty!(v::IndexedSet)
+    nzidx = v.nzidx
+    empty = v.empty
+    for i in 1:v.nnz
+        empty[nzidx[i]] = true
+    end
+    v.nnz = 0
+    return v
+end
+
+Base.length(v::IndexedSet) = length(v.nzidx)
+
+function Base.resize!(v::IndexedSet, n::Integer)
+    if n > length(v)
+        @assert v.nnz == 0 # only resize empty vector
+        resize!(v.nzidx, n)
+        resize!(v.empty, n)
+        fill!(v.empty, true)
+    end
+    return
+end
+
+Base.collect(v::IndexedSet) = v.nzidx[1:v.nnz]
+
+function Base.union!(v::IndexedSet, s)
+    for x in s
+        push!(v, x)
+    end
+    return
+end
+
+# compact storage for an undirected graph
+# neighbors of vertex i start at adjlist[offsets[i]]
+struct UndirectedGraph
+    adjlist::Vector{Int}
+    edgeindex::Vector{Int} # corresponding edge number, indexed as adjlist
+    offsets::Vector{Int}
+    edges::Vector{Tuple{Int,Int}}
+end
+
+_num_vertices(g::UndirectedGraph) = length(g.offsets) - 1
+
+_num_edges(g::UndirectedGraph) = length(g.edges)
+
+_num_neighbors(i::Int, g::UndirectedGraph) = g.offsets[i+1] - g.offsets[i]
+
+_start_neighbors(i::Int, g::UndirectedGraph) = g.offsets[i]
+
+function gen_adjlist(I, J, nel)
+    adjcount = zeros(Int, nel)
+    n_edges = 0
+    for k in 1:length(I)
+        i = I[k]
+        j = J[k]
+        if i == j
+            continue
+        end
+        n_edges += 1
+        adjcount[i] += 1
+        adjcount[j] += 1
+    end
+    offsets = Array{Int}(undef, nel + 1)
+    offsets[1] = 1
+    for k in 1:nel
+        offsets[k+1] = offsets[k] + adjcount[k]
+    end
+    fill!(adjcount, 0)
+    edges = Array{Tuple{Int,Int}}(undef, n_edges)
+    adjlist = Array{Int}(undef, offsets[nel+1] - 1)
+    edgeindex = Array{Int}(undef, length(adjlist))
+    edge_count = 0
+    for k in 1:length(I)
+        i = I[k]
+        j = J[k]
+        if i == j
+            continue
+        end
+        edge_count += 1
+        adjlist[offsets[i]+adjcount[i]] = j
+        edgeindex[offsets[i]+adjcount[i]] = edge_count
+        adjcount[i] += 1
+        adjlist[offsets[j]+adjcount[j]] = i
+        edgeindex[offsets[j]+adjcount[j]] = edge_count
+        adjcount[j] += 1
+        edges[edge_count] = (i, j)
+    end
+    @assert edge_count == n_edges
+    return UndirectedGraph(adjlist, edgeindex, offsets, edges)
+end
+
+struct _Edge
```

Comment:
If this is a simple and not hypergraph, an index information is not necessary, as the edge is fully described by src and dst?

Reply:
also, if the graph is part of the public interface, it would be weird for the Edge type not to be

Reply:
Here and the rest of `Coloring.jl`: I avoided making any major changes. The goal of this was to tidy and refactor the existing implementation, not to improve it.